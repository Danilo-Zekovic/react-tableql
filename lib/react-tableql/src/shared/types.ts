export interface TableQLThemeAdvanced {
  header?: {
    color?: string
    background?: string
  }
  rows?: {
    color?: string
    background?: string
    hover?: string
    stripes?: {
      color?: string
      background?: string
      hover?: string
    }
  }
  grid?: {
    horizontal?: string
    vertical?: string
    frame?: string
  }
}

export interface TableQLThemeBasic {
  primary?: string // main color for the table, header and colored striped rows
  secondary?: string // rows that are not colored in striped scenario or all rows in non striped setup
  primaryFontColor?: string // font color for header and striped rows
  secondaryFontColor?: string // color for rows
  striped?: boolean // should the table rows be striped
  grid?: string // grid line color
}

export type TableQLTheme = TableQLThemeAdvanced | TableQLThemeBasic

export type Data = object[] | object | []
export interface ColumnConfig {
  id: string
  label?: string
  component?: (data: unknown) => void
  customColumn?: boolean
  headerStyle?: string
  nodeStyle?: string | ((data: unknown) => void)
  sort?: boolean | ((newOrder: unknown, property: string) => void)
}

export type Column = string | ColumnConfig
