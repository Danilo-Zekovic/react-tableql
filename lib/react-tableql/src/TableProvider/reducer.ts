import { Action, State } from './types'

export const tableReducer = (
  state: State,
  { type, payload }: Action,
): State => {
  // if (!theme) {
  //   return { theme: { ...state.theme } }
  // }

  switch (type) {
    case 'setTheme': {
      if (!payload.theme) {
        // TODO consider maybe throwing error
        // throw new Error(`Theme not passed.`)
        return { ...state }
      }
      return { ...state, theme: { ...payload.theme } }
    }
    case 'setData': {
      return { ...state, data: [...payload.data] }
    }

    case 'setDataKeys': {
      return { ...state, dataKeys: [...payload.dataKeys] }
    }
    default: {
      throw new Error(`Unhandled action type: ${type}`)
    }
  }
}
