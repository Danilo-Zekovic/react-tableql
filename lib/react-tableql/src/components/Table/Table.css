table {
  border: 1px solid;
  border-collapse: collapse;
  margin: 0.5rem 0;
  padding: 0 0.4rem;
  width: 100%;
  table-layout: fixed;
}

table caption {
  font-size: 1.5rem;
  margin: 0.5rem 0 0.5rem;
}

table tr {
  background-color: 'transparent'; /* #f8f8f8;*/
  border: 1px solid #ddd;
  /* padding: 0.35em; */
}

table th,
table td {
  padding: 0.69rem;
  text-align: center;
}

table th {
  font-size: 0.85rem;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  border-bottom: 2px solid #dee2e6;
}

@media screen and (max-width: 740px) {
  table {
    border: 0;
  }

  table caption {
    font-size: 1.3rem;
  }

  table thead {
    border: none;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }

  table tr {
    border-bottom: 3px solid #ddd;
    display: block;
    margin-bottom: 0.69rem;
  }

  table td {
    border-bottom: 1px solid #ddd;
    display: block;
    font-size: 0.8rem;
    text-align: right;
  }

  table td::before {
    /*
    * aria-label has no advantage, it won't be read inside a table
    content: attr(aria-label);
    */
    content: attr(data-label);
    float: left;
    font-weight: bold;
    text-transform: uppercase;
  }

  table td:last-child {
    border-bottom: 0;
  }
}
