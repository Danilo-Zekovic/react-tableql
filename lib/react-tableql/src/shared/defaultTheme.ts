import { TableQLProviderTheme } from '../TableQLProvider/types'

export const defaultTheme: TableQLProviderTheme = {
  header: {
    color: '#000',
    background: 'transparent', //'#fff',
  },
  rows: {
    color: '#000',
    background: 'transparent', //'#fff',
    hover: '#f1f3f5',
    stripes: {
      color: '#000',
      background: 'transparent', //'#fff',
      hover: '#f1f3f5',
    },
  },
  grid: {
    horizontal: '#dee2e6',
    vertical: '#dee2e6',
    frame: '#dee2e6',
  },
}
