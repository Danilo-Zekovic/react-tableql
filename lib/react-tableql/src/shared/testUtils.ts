export const MOCK_DATA = [
  {
    author: 'Austen, Jane',
    title: 'Northanger Abbey',
    id: '63a32284072b2e31190bd81d',
    edition: 'Penguin',
    price: 18.2,
    yearWritten: 1814,
  },
  {
    author: 'Tolstoy, Leo',
    title: 'War and Peace',
    id: '63a32284072b2e31190bd81e',
    edition: 'Penguin',
    price: 12.7,
    yearWritten: 1865,
  },
  {
    author: 'Tolstoy, Leo',
    title: 'Anna Karenina',
    id: '63a32284072b2e31190bd81f',
    edition: 'Penguin',
    price: 13.5,
    yearWritten: 1875,
  },
  {
    author: 'Woolf, Virginia',
    title: 'Mrs. Dalloway',
    id: '63a32284072b2e31190bd820',
    edition: 'Harcourt Brace',
    price: 25,
    yearWritten: 1925,
  },
  {
    author: 'Cunnningham, Michael',
    title: 'The Hours',
    id: '63a32284072b2e31190bd821',
    edition: 'Harcourt Brace',
    price: 12.35,
    yearWritten: 1999,
  },
  {
    author: 'Twain, Mark',
    title: 'Huckleberry Finn',
    id: '63a32284072b2e31190bd822',
    edition: 'Penguin',
    price: 5.76,
    yearWritten: 1865,
  },
  {
    author: 'Dickens, Charles',
    title: 'Bleak House',
    id: '63a32284072b2e31190bd823',
    edition: 'Random House',
    price: 5.75,
    yearWritten: 1870,
  },
  {
    author: 'Twain, Mark',
    title: 'Tom Sawyer',
    id: '63a32284072b2e31190bd824',
    edition: 'Random House',
    price: 7.75,
    yearWritten: 1862,
  },
  {
    author: 'Woolf, Virginia',
    title: "A Room of One's Own",
    id: '63a32284072b2e31190bd825',
    edition: 'Penguin',
    price: 29,
    yearWritten: 1922,
  },
  {
    author: 'Rowling, J.K.',
    title: 'Harry Potter',
    id: '63a32284072b2e31190bd826',
    edition: 'Harcourt Brace',
    price: 19.95,
    yearWritten: 2000,
  },
  {
    author: 'Marquez',
    title: 'One Hundred Years of Solitude',
    id: '63a32284072b2e31190bd827',
    edition: 'Harper  Perennial',
    price: 14,
    yearWritten: 1967,
  },
  {
    author: 'Shakespeare',
    title: 'Hamlet, Prince of Denmark',
    id: '63a32284072b2e31190bd828',
    edition: 'Signet  Classics',
    price: 7.95,
    yearWritten: 1603,
  },
  {
    author: 'Tolkien, J.R.',
    title: 'Lord of the Rings',
    id: '63a32284072b2e31190bd829',
    edition: 'Penguin',
    price: 27.45,
    yearWritten: 1937,
  },
]

export const MOCK_DATA_KEYS = Object.keys(MOCK_DATA[0])
