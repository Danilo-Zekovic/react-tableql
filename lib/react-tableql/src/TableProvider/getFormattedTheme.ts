import { defaultTheme } from '../shared/defaultTheme'
import {
  TableQLTheme,
  TableQLThemeAdvanced,
  TableQLThemeBasic,
} from '../shared/types'
import { TableQLProviderTheme } from '../TableQLProvider/types'

const isBasicTheme = (theme: TableQLTheme): theme is TableQLThemeBasic => {
  if (
    (theme as TableQLThemeBasic).primary ||
    (theme as TableQLThemeBasic).secondary ||
    (theme as TableQLThemeBasic).primaryFontColor ||
    (theme as TableQLThemeBasic).secondaryFontColor ||
    (theme as TableQLThemeBasic).striped ||
    (theme as TableQLThemeBasic).primary
  ) {
    return true
  }
  return false
}

const formattedBasicTheme = ({
  primary,
  primaryFontColor,
  secondaryFontColor,
  secondary,
  grid,
}: TableQLThemeBasic): TableQLProviderTheme => {
  const {
    header,
    rows,
    grid: { horizontal, vertical },
  } = defaultTheme

  return {
    header: {
      color: primaryFontColor || header.color,
      background: primary || header.background,
    },
    rows: {
      color: secondaryFontColor || rows.color,
      background: secondary || rows.background,
      hover: rows.hover,
      stripes: {
        color: primaryFontColor || rows.stripes.color,
        background: primaryFontColor || rows.stripes.background,
        hover: rows.stripes.hover,
      },
    },
    grid: {
      horizontal: grid || horizontal,
      vertical: grid || vertical,
    },
  }
}

const formattedAdvancedTheme = ({
  header,
  rows,
  grid,
}: TableQLThemeAdvanced): TableQLProviderTheme => {
  return {
    header: { ...defaultTheme.header, ...header },
    rows: {
      ...defaultTheme.rows,
      ...rows,
      stripes: {
        ...defaultTheme.rows.stripes,
        ...rows?.stripes,
      },
    },
    grid: { ...defaultTheme.grid, ...grid },
  }
}

export const getFormattedTheme = (
  theme: TableQLTheme,
): TableQLProviderTheme => {
  return isBasicTheme(theme)
    ? formattedBasicTheme(theme)
    : formattedAdvancedTheme(theme)
}
