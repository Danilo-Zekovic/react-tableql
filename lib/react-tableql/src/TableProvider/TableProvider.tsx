import React, {
  createContext,
  FC,
  ReactElement,
  useContext,
  useReducer,
} from 'react'

import { defaultTheme } from '../shared/defaultTheme'
import { useTableQLState } from '../TableQLProvider'
import { getFormattedTheme } from './getFormattedTheme'
import { tableReducer } from './reducer'
import { State, TableProviderProps } from './types'

const TableStateContext = createContext<State>({ theme: defaultTheme })

const TableProvider: FC<TableProviderProps> = ({
  children,
  theme,
}): ReactElement => {
  // TODO as of right now this is line makes it mandatory to have the TableQLProvider, but we do not want to force it onto users
  const { theme: globalTheme } = useTableQLState()
  const [state] = useReducer(tableReducer, {
    // if there is no local theme being passed use the global
    // TODO this should probably be improved so that if global is not set either just don't use it
    theme: getFormattedTheme(theme || globalTheme),
  })

  return (
    <TableStateContext.Provider value={state}>
      {children}
    </TableStateContext.Provider>
  )
}

const useTableState = (): State => {
  const context = useContext(TableStateContext)
  /* istanbul ignore next */
  if (context === undefined) {
    throw new Error('useTableState must be used within a TableQLProvider')
  }
  return context
}

export { TableProvider, useTableState }
