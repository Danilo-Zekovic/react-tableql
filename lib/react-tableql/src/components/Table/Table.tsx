import React, { FC } from 'react'
import { useTableState } from '../../TableProvider'
import { TableHeader } from '../TableHeader'
import { TableRows } from '../TableRows'

import './Table.css'

const Table: FC = () => {
  const { data, dataKeys, onEmpty } = useTableState()

  if (Array.isArray(data) && data.length === 0) {
    // log('No data found!', data.length)
    if (onEmpty) {
      return onEmpty({ reason: 'No data found!' })
    }

    // if no data, and no columns defined
    if (!dataKeys.length) {
      return <p>{`No data found!`}</p>
    }
  }
  return (
    <table className={`TableQL`}>
      {/* TODO caption should be customizable */}
      <caption className={`TableQL-caption`}>Statement Summary</caption>
      <thead className={`TableQL-thead`}>
        <TableHeader />
      </thead>
      <tbody className={`TableQL-tbody`}>
        <TableRows />
      </tbody>
    </table>
  )
}

export default Table
