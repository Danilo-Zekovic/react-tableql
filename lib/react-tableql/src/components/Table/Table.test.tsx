import React, { FC } from 'react'

import { render, screen } from '@testing-library/react'

import Table from './Table'
import { TableProvider } from '../../TableProvider'
import { MOCK_DATA } from '../../shared/testUtils'

describe('<Table>', () => {
  it('default snapshot', () => {
    const { container } = render(
      <TableProvider data={MOCK_DATA}>
        <Table />)
      </TableProvider>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('pass onEmpty', () => {
    const Empty: FC = () => <h3>I am empty</h3>
    const { container } = render(
      <TableProvider data={[]} onEmpty={Empty}>
        <Table />)
      </TableProvider>,
    )
    expect(screen.getByText('I am empty')).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })
})
