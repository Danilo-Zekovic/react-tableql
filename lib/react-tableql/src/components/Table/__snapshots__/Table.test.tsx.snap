// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table> default snapshot 1`] = `
<table
  class="TableQL"
>
  <caption
    class="TableQL-caption"
  >
    Statement Summary
  </caption>
  <thead
    class="TableQL-thead"
  >
    <tr
      class="TableQL-tr"
    >
      <th
        class="TableQL-th"
        scope="col"
      >
        Author
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Title
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Id
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Edition
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Price
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Year Written
      </th>
    </tr>
  </thead>
  <tbody
    class="TableQL-tbody"
  >
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Austen, Jane
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Northanger Abbey
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd81d
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        18.2
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1814
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Tolstoy, Leo
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        War and Peace
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd81e
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        12.7
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1865
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Tolstoy, Leo
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Anna Karenina
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd81f
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        13.5
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1875
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Woolf, Virginia
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Mrs. Dalloway
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd820
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Harcourt Brace
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        25
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1925
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Cunnningham, Michael
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        The Hours
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd821
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Harcourt Brace
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        12.35
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1999
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Twain, Mark
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Huckleberry Finn
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd822
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        5.76
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1865
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Dickens, Charles
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Bleak House
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd823
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Random House
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        5.75
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1870
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Twain, Mark
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Tom Sawyer
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd824
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Random House
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        7.75
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1862
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Woolf, Virginia
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        A Room of One's Own
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd825
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        29
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1922
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Rowling, J.K.
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Harry Potter
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd826
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Harcourt Brace
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        19.95
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        2000
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Marquez
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        One Hundred Years of Solitude
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd827
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Harper  Perennial
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        14
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1967
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Shakespeare
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Hamlet, Prince of Denmark
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd828
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Signet  Classics
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        7.95
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1603
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="author"
      >
        Tolkien, J.R.
      </td>
      <td
        class="TableQL-td"
        data-label="title"
      >
        Lord of the Rings
      </td>
      <td
        class="TableQL-td"
        data-label="id"
      >
        63a32284072b2e31190bd829
      </td>
      <td
        class="TableQL-td"
        data-label="edition"
      >
        Penguin
      </td>
      <td
        class="TableQL-td"
        data-label="price"
      >
        27.45
      </td>
      <td
        class="TableQL-td"
        data-label="yearWritten"
      >
        1937
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> pass onEmpty 1`] = `
<h3>
  I am empty
</h3>
`;
